apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/bulk/S3BucketConnection.scala

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package org.lfdecentralizedtrust.splice.scan.store.bulk
+
+import com.digitalasset.canton.logging.{NamedLoggerFactory, NamedLogging}
+import com.digitalasset.canton.tracing.TraceContext
+import software.amazon.awssdk.auth.credentials.{AwsBasicCredentials, StaticCredentialsProvider}
+import software.amazon.awssdk.core.async.{AsyncRequestBody, AsyncResponseTransformer}
+import software.amazon.awssdk.regions.Region
+import software.amazon.awssdk.services.s3.{S3AsyncClient, S3Configuration}
+import software.amazon.awssdk.services.s3.model.{
+  GetObjectRequest,
+  GetObjectResponse,
+  PutObjectRequest,
+}
+
+import scala.jdk.FutureConverters.*
+import java.net.URI
+import java.nio.ByteBuffer
+import scala.concurrent.{ExecutionContext, Future}
+
+case class S3Config(
+    endpoint: URI,
+    bucketName: String,
+    region: Region,
+    credentials: AwsBasicCredentials,
+)
+
+class S3BucketConnection(
+    val s3Client: S3AsyncClient,
+    val bucketName: String,
+    val loggerFactory: NamedLoggerFactory,
+) extends NamedLogging {
+  // Reads the full content of an s3 object into a ByteBuffer.
+  // Use only for testing, when the object size is known to be small
+  def readFullObject(key: String)(implicit ec: ExecutionContext): Future[ByteBuffer] = {
+    val request = GetObjectRequest.builder().bucket(bucketName).key(key).build()
+    s3Client
+      .getObject(request, AsyncResponseTransformer.toBytes[GetObjectResponse])
+      .asScala
+      .map(_.asByteBuffer())
+  }
+
+  // Writes a full object from memory into an s3 object
+  def writeFullObject(key: String, content: ByteBuffer)(implicit
+      tc: TraceContext,
+      ec: ExecutionContext,
+  ): Future[Unit] = {
+    logger.debug(s"Writing ${content.array().length} bytes to S3 object $key")
+    val putObj: PutObjectRequest = PutObjectRequest
+      .builder()
+      .bucket(bucketName)
+      .key(key)
+      .build()
+    s3Client
+      .putObject(
+        putObj,
+        AsyncRequestBody.fromBytes(content.array()),
+      )
+      .asScala
+      .map(_ => ())
```

Comment:
Maybe at a later stage you can get the result and check ETag against md5 hash of what you wrote, or something like that to guarantee it's really there. (not meant now in this PR)